ktor-server/ktor-server-config-yaml/jvmAndNix/src/io/ktor/server/config/yaml/YamlConfig.kt

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2014-2022 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+package io.ktor.server.config.yaml
+
+import io.ktor.server.config.*
+import net.mamoe.yamlkt.*
+
+public class YamlConfigLoader : ConfigLoader {
+    override fun load(path: String?): ApplicationConfig? {
+        return YamlConfig(path)?.apply { checkEnvironmentVariables() }
+    }
+}
+
+/**
+ * Loads config from YAML file if found.
+ * On JVM uses resource if exists, otherwise reads from file.
+ * On Native always reads from file.
+ */
+public expect fun YamlConfig(path: String?): YamlConfig?
+
+/**
+ * Implements [ApplicationConfig] by loading configuration from YAML file.
+ * Values can reference to environment variables with `$ENV_VAR` syntax.
+ */
+public class YamlConfig(private val yaml: YamlMap) : ApplicationConfig {
+
+    override fun property(path: String): ApplicationConfigValue {
+        return propertyOrNull(path) ?: throw ApplicationConfigurationException("Path $path not found.")
+    }
+
+    override fun propertyOrNull(path: String): ApplicationConfigValue? {
+        val parts = path.split('.')
+        val yaml = parts.dropLast(1).fold(yaml) { yaml, part -> yaml[part] as? YamlMap ?: return null }
+        val value = yaml[parts.last()] ?: return null
+        return ConfigValue(value, path)
+    }
+
+    override fun config(path: String): ApplicationConfig {
+        val parts = path.split('.')
+        val yaml = parts.fold(yaml) { yaml, part ->
+            yaml[part] as? YamlMap ?: throw ApplicationConfigurationException("Path $path not found.")
+        }
+        return YamlConfig(yaml)
+    }
+
+    override fun configList(path: String): List<ApplicationConfig> {
+        val parts = path.split('.')
+        val yaml = parts.dropLast(1).fold(yaml) { yaml, part ->
+            yaml[part] as? YamlMap ?: throw ApplicationConfigurationException("Path $path not found.")
+        }
+        val value = yaml[parts.last()] as? YamlList ?: throw ApplicationConfigurationException("Path $path not found.")
+        return value.map {
+            YamlConfig(
+                it as? YamlMap
+                    ?: throw ApplicationConfigurationException("Property $path is not a list of maps.")
+            )
+        }
+    }
+
+    override fun keys(): Set<String> {
+        fun keys(yaml: YamlMap): Set<String> {
```

Comment:
Could you tell me why do we need an internal function?

Reply:
no real need, but I prefer to use local internal functions if they are helpers for one specific external ones

Reply:
consider extracting such functions to the top level to make it static and prevent capturing `this`